
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <meta name="HandheldFriendly" content="True" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta name="robots" content="" />

  <link href="https://fonts.googleapis.com/css?family=Source+Code+Pro|Source+Sans+Pro:300,400,400i,700" rel="stylesheet">

    <link rel="stylesheet" type="text/css" href="/theme/stylesheet/style.min.css">

  <link rel="stylesheet" type="text/css" href="/theme/pygments/github.min.css">
  <link rel="stylesheet" type="text/css" href="/theme/font-awesome/css/font-awesome.min.css">







<meta name="author" content="Colm Ryan" />
<meta name="description" content="I had to replace a CLC Bio variant calling pipeline with an open-source equivalent. These notes are mainly an update of another blog post Variant calling with GATK. Tools Setup Build and install (to ~/.local) bwa Build and install samtools Download and extract GATK under non-comercial license Download and extract …" />
<meta name="keywords" content="Julia">

<meta property="og:site_name" content="Tomorrow Said Toad"/>
<meta property="og:title" content="Variant calls with GATK"/>
<meta property="og:description" content="I had to replace a CLC Bio variant calling pipeline with an open-source equivalent. These notes are mainly an update of another blog post Variant calling with GATK. Tools Setup Build and install (to ~/.local) bwa Build and install samtools Download and extract GATK under non-comercial license Download and extract …"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:url" content="/variant-calls-with-gatk.html"/>
<meta property="og:type" content="article"/>
<meta property="article:published_time" content="2016-05-29 00:00:00-04:00"/>
<meta property="article:modified_time" content="2016-05-29 00:00:00-04:00"/>
<meta property="article:author" content="/author/colm-ryan.html">
<meta property="article:section" content="Bioinformatics"/>
<meta property="article:tag" content="Julia"/>
<meta property="og:image" content="">

  <title>Tomorrow Said Toad &ndash; Variant calls with GATK</title>

</head>
<body>
  <aside>
    <div>
      <a href="">
        <img src="/theme/img/profile.png" alt="Tomorrow Said Toad" title="Tomorrow Said Toad">
      </a>
      <h1><a href="">Tomorrow Said Toad</a></h1>

<p>Today I will take life easy...</p>
      <nav>
        <ul class="list">

          <li><a href="http://getpelican.com/" target="_blank">Pelican</a></li>
          <li><a href="http://python.org/" target="_blank">Python.org</a></li>
        </ul>
      </nav>

      <ul class="social">
        <li><a class="sc-github" href="http://github.com/caryan" target="_blank"><i class="fa fa-github"></i></a></li>
      </ul>
    </div>


  </aside>
  <main>


<article class="single">
  <header>
      
    <h1 id="variant-calls-with-gatk">Variant calls with GATK</h1>
    <p>
          Posted on Sun 29 May 2016 in <a href="/category/bioinformatics.html">Bioinformatics</a>


    </p>
  </header>


  <div>
    <p>I had to replace a <a href="http://www.clcbio.com/">CLC Bio</a> variant calling pipeline
with an open-source equivalent. These notes are mainly an update of another blog
post <a href="https://approachedinthelimit.wordpress.com/2015/10/09/variant-calling-with-gatk/">Variant calling with
GATK</a>.</p>
<h1>Tools Setup</h1>
<ol>
<li>Build and install (to ~/.local) <a href="https://github.com/lh3/bwa">bwa</a></li>
<li>Build and install <a href="http://www.htslib.org/download/">samtools</a></li>
<li>Download and extract <a href="https://www.broadinstitute.org/gatk/">GATK</a> under non-comercial license</li>
<li>Download and extract <a href="http://broadinstitute.github.io/picard/">Picard Tools</a></li>
<li>Picard requires Java 1.8 so update to OpenJDK-1.8.72</li>
<li>Set environment variable to reference Picard<div class="highlight"><pre><span></span><span class="nb">export</span> <span class="nv">PICARD</span><span class="o">=</span>~/Downloads/picard-tools-2.1.0/picard.jar
</pre></div>


</li>
</ol>
<h1>Pipeline</h1>
<ol>
<li>
<p>Index the reference</p>
<div class="highlight"><pre><span></span>bwa index PAO1_reference.fasta
</pre></div>


</li>
<li>
<p>Sort the reference</p>
<div class="highlight"><pre><span></span>samtools faidx PAO1_reference.fasta
</pre></div>


</li>
<li>
<p>Create sequence dictionary</p>
<div class="highlight"><pre><span></span>java -jar <span class="nv">$PICARD</span> CreateSequenceDictionary <span class="se">\</span>
 <span class="nv">REFERENCE</span><span class="o">=</span>PAO1_reference.fasta <span class="se">\</span>
 <span class="nv">OUTPUT</span><span class="o">=</span>PAO1_reference.dict
</pre></div>


</li>
<li>
<p>Align reads using bwa</p>
<div class="highlight"><pre><span></span>bwa mem PAO1_reference.fasta <span class="se">\</span>
 genomes/AZPAE12150_S4_L001_R1_001.fastq.gz <span class="se">\</span>
 genomes/AZPAE12150_S4_L001_R2_001.fastq.gz &gt; <span class="se">\</span>
 pipeline_products/AZPAE12150_S4_L001.aln.sam
</pre></div>


</li>
<li>
<p>Sort SAM file, add fake ReadGroup IDs (otherwise will fail at GATK RealignerTargetCreator below) and convert to bam</p>
<div class="highlight"><pre><span></span>java -jar <span class="nv">$PICARD</span> AddOrReplaceReadGroups <span class="se">\</span>
 <span class="nv">I</span><span class="o">=</span>pipeline_products/AZPAE12150_S4_L001.aln.sam <span class="se">\</span>
 <span class="nv">O</span><span class="o">=</span>pipeline_products/AZPAE12150_S4_L001.sorted.bam <span class="se">\</span>
 <span class="nv">RGLB</span><span class="o">=</span>kos1 <span class="se">\</span>
 <span class="nv">RGPL</span><span class="o">=</span>illumina <span class="se">\</span>
 <span class="nv">RGPU</span><span class="o">=</span>unit1 <span class="se">\</span>
 <span class="nv">RGSM</span><span class="o">=</span>AZPAE12150 <span class="se">\</span>
 <span class="nv">SORT_ORDER</span><span class="o">=</span>coordinate
</pre></div>


</li>
<li>
<p>Mark duplicates</p>
<div class="highlight"><pre><span></span>java -jar <span class="nv">$PICARD</span> MarkDuplicates <span class="se">\</span>
 <span class="nv">I</span><span class="o">=</span>pipeline_products/AZPAE12150_S4_L001.sorted.bam<span class="se">\</span>
 <span class="nv">O</span><span class="o">=</span>pipeline_products/AZPAE12150_S4_L001.dedup.bam <span class="se">\</span>
 <span class="nv">METRICS_FILE</span><span class="o">=</span>pipeline_products/metrics.txt
</pre></div>


</li>
<li>
<p>Build bam index</p>
<div class="highlight"><pre><span></span>java -jar <span class="nv">$PICARD</span> BuildBamIndex <span class="nv">INPUT</span><span class="o">=</span>pipeline_products/AZPAE12150_S4_L001.dedup.bam
</pre></div>


</li>
<li>
<p>Create realignment targets</p>
<div class="highlight"><pre><span></span>java -jar ~/Downloads/GATK/GenomeAnalysisTK.jar -T RealignerTargetCreator <span class="se">\</span>
-R ../PAO1_reference.fasta -I AZPAE12150_S4_L001.dedup.bam <span class="se">\</span>
-o AZPAE12150_S4_L001.targetintervals.list
</pre></div>


</li>
<li>
<p>Indel realignment</p>
<div class="highlight"><pre><span></span>java -jar ~/Downloads/GATK/GenomeAnalysisTK.jar -T IndelRealigner <span class="se">\</span>
 -R ../PAO1_reference.fasta -I AZPAE12150_S4_L001.dedup.bam <span class="se">\</span>
 -targetIntervals AZPAE12150_S4_L001.targetintervals.list <span class="se">\</span>
 -o AZPAE12150_S4_L001.realigned.bam
</pre></div>


</li>
<li>
<p>Call variants (takes ~10 minutes to run on a pair of 150M fastq.gz file )</p>
<div class="highlight"><pre><span></span>java -jar ~/Downloads/GATK/GenomeAnalysisTK.jar -T HaplotypeCaller <span class="se">\</span>
 -R ../PAO1_reference.fasta -I AZPAE12150_S4_L001.realigned.bam <span class="se">\</span>
 -ploidy <span class="m">1</span> -stand_call_conf <span class="m">30</span> -stand_emit_conf <span class="m">10</span> <span class="se">\</span>
 -o AZPAE12150_S4_L001.raw.vcf
</pre></div>


</li>
<li>
<p>Filter out SNPs (optional)</p>
<div class="highlight"><pre><span></span>java -jar ~/Downloads/GATK/GenomeAnalysisTK.jar -T SelectVariants <span class="se">\</span>
 -R ../PAO1_reference.fasta -V AZPAE12150_S4_L001.raw.vcf -selectType SNP <span class="se">\</span>
 -o AZPAE12150_S4_L001.snps.vcf
</pre></div>


</li>
<li>
<p>Convert to a tab separated file for easy dataframe loading</p>
<div class="highlight"><pre><span></span>java -jar ~/Downloads/GATK/GenomeAnalysisTK.jar -T VariantsToTable <span class="se">\</span>
 -R ../PAO1_reference.fasta -V AZPAE12150_S4_L001.snps.vcf <span class="se">\</span>
 -F POS -F REF -F ALT -F QUAL -o AZPAE12150_S4_L001.snps.tsv
</pre></div>


</li>
</ol>
<h1>Wrapping in a script</h1>
<p>I also put together a <a href="https://gist.github.com/caryan/1a5624d8539c83d01cef2663b3bc2e7d">Julia
script</a> that
automates the whole pipeline.</p>
<h1>Performance</h1>
<p>I don't have the experience to grade the quality of the SNP calling. Anecdotally
I seemed to run into a <a href="http://gatkforums.broadinstitute.org/dsde/discussion/5422/why-is-haplotypecaller-not-calling-obvious-snps-by-default-but-requires-allownonuniquekmersinref">reported
issue</a>
where it seemed to miss SNPs at the beginning of a gene with lots of repeats.</p>
<p>In comparison to the CLC pipeline, some SNP set union/differences showed there
were about 27k they both called;  CLC called a few hundred that GATK did not and
GATK called a couple thousand that CLC did not.  Anecdotally, the extra calls
from GATK seemed to come from low depth of coverage region however some
primitive attempts to filter those using <code>SelectVariants</code> did not seem to
improve the overlap.</p>
  </div>
  <div class="tag-cloud">
    <p>
      <a href="/tag/julia.html">Julia</a>
    </p>
  </div>





</article>

    <footer>
<p>
  &copy;  2019 - This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>
</p>
<p>    Powered by <a href="http://getpelican.com" target="_blank">Pelican</a> - <a href="https://github.com/alexandrevicenzi/flex" target="_blank">Flex</a> theme by <a href="http://alexandrevicenzi.com" target="_blank">Alexandre Vicenzi</a>
</p><p>
  <a rel="license"
     href="http://creativecommons.org/licenses/by-sa/4.0/"
     target="_blank">
    <img alt="Creative Commons License"
         title="Creative Commons License"
         style="border-width:0"
           src="https://i.creativecommons.org/l/by-sa/4.0/80x15.png"
         width="80"
         height="15"/>
  </a>
</p>    </footer>
  </main>




<script type="application/ld+json">
{
  "@context" : "http://schema.org",
  "@type" : "Blog",
  "name": " Tomorrow Said Toad ",
  "url" : "",
  "image": "",
  "description": ""
}
</script>

</body>
</html>